File: haskell/src/Wildling/Json.hs
module Wildling.Json
  ( JsonValue(..)
  , parseJson
  ) where

import Data.Char (chr, isDigit, isHexDigit, isSpace)
import Data.List (foldl')

data JsonValue
  = JsonNull
  | JsonBool Bool
  | JsonNumber Double
  | JsonString String
  | JsonArray [JsonValue]
  | JsonObject [(String, JsonValue)]
  deriving (Eq, Show)

parseJson :: String -> Either String JsonValue
parseJson text =
  case parseValue (dropWhile isSpace text) of
    Left err -> Left err
    Right (value, rest) ->
      case dropWhile isSpace rest of
        "" -> Right value
        _ -> Left "Trailing content"

parseValue :: String -> Either String (JsonValue, String)
parseValue s =
  case dropWhile isSpace s of
    [] -> Left "Unexpected end"
    '{' : rest -> parseObject rest
    '[' : rest -> parseArray rest
    '"' : rest -> do
      (str, rest') <- parseString rest
      pure (JsonString str, rest')
    't' : 'r' : 'u' : 'e' : rest -> pure (JsonBool True, rest)
    'f' : 'a' : 'l' : 's' : 'e' : rest -> pure (JsonBool False, rest)
    'n' : 'u' : 'l' : 'l' : rest -> pure (JsonNull, rest)
    c : _
      | c == '-' || isDigit c -> parseNumber s'
      | otherwise -> Left "Unexpected character"
      where
        s' = dropWhile isSpace s

parseObject :: String -> Either String (JsonValue, String)
parseObject s0 =
  let s = dropWhile isSpace s0
   in case s of
        '}' : rest -> pure (JsonObject [], rest)
        _ -> go [] s
  where
    go acc s =
      case dropWhile isSpace s of
        '"' : rest -> do
          (key, rest1) <- parseString rest
          case dropWhile isSpace rest1 of
            ':' : rest2 -> do
              (value, rest3) <- parseValue rest2
              let acc' = acc ++ [(key, value)]
              case dropWhile isSpace rest3 of
                '}' : rest4 -> pure (JsonObject acc', rest4)
                ',' : rest4 -> go acc' rest4
                _ -> Left "Expected ',' or '}' in object"
            _ -> Left "Expected ':' in object"
        _ -> Left "Expected string key in object"

parseArray :: String -> Either String (JsonValue, String)
parseArray s0 =
  let s = dropWhile isSpace s0
   in case s of
        ']' : rest -> pure (JsonArray [], rest)
        _ -> go [] s
  where
    go acc s = do
      (value, rest1) <- parseValue s
      let acc' = acc ++ [value]
      case dropWhile isSpace rest1 of
        ']' : rest2 -> pure (JsonArray acc', rest2)
        ',' : rest2 -> go acc' rest2
        _ -> Left "Expected ',' or ']' in array"

parseString :: String -> Either String (String, String)
parseString = go []
  where
    go acc [] = Left "Unterminated string"
    go acc ('"' : rest) = pure (reverse acc, rest)
    go acc ('\\' : rest) = do
      (c, rest') <- parseEscape rest
      go (c : acc) rest'
    go acc (c : rest) = go (c : acc) rest

parseEscape :: String -> Either String (Char, String)
parseEscape [] = Left "Unterminated escape"
parseEscape ('"' : rest) = pure ('"', rest)
parseEscape ('\\' : rest) = pure ('\\', rest)
parseEscape ('/' : rest) = pure ('/', rest)
parseEscape ('b' : rest) = pure ('\b', rest)
parseEscape ('f' : rest) = pure ('\f', rest)
parseEscape ('n' : rest) = pure ('\n', rest)
parseEscape ('r' : rest) = pure ('\r', rest)
parseEscape ('t' : rest) = pure ('\t', rest)
parseEscape ('u' : a : b : c : d : rest)
  | all isHexDigit [a, b, c, d] =
      let code = foldl' (\n x -> n * 16 + hexVal x) 0 [a, b, c, d]
       in pure (chr code, rest)
  | otherwise = Left "Invalid unicode escape"
parseEscape _ = Left "Invalid escape"

hexVal :: Char -> Int
hexVal c
  | c >= '0' && c <= '9' = fromEnum c - fromEnum '0'
  | c >= 'a' && c <= 'f' = 10 + fromEnum c - fromEnum 'a'
  | c >= 'A' && c <= 'F' = 10 + fromEnum c - fromEnum 'A'
  | otherwise = 0

parseNumber :: String -> Either String (JsonValue, String)
parseNumber s =
  let (raw, rest, isFloat) = takeNumber s
   in if null raw
        then Left "Invalid number"
        else
          if isFloat
            then case reads raw :: [(Double, String)] of
              [(n, "")] -> pure (JsonNumber n, rest)
              _ -> Left "Invalid number"
            else case reads raw :: [(Integer, String)] of
              [(n, "")] -> pure (JsonNumber (fromIntegral n), rest)
              _ -> Left "Invalid number"

takeNumber :: String -> (String, String, Bool)
takeNumber s = go s [] False
  where
    go ('-' : rest) acc flag = go rest ('-' : acc) flag
    go (c : rest) acc flag
      | isDigit c = go rest (c : acc) flag
    go ('.' : rest) acc _ =
      let (digits, rest', _) = takeDigits rest
       in continueExp ('.' : digits ++ acc) rest' True
    go rest acc flag = continueExp acc rest flag

    continueExp acc ('e' : rest) _ = takeExp ('e' : acc) rest True
    continueExp acc ('E' : rest) _ = takeExp ('E' : acc) rest True
    continueExp acc rest flag = (reverse acc, rest, flag)

    takeExp acc ('+' : rest) flag = takeExpDigits ('+' : acc) rest flag
    takeExp acc ('-' : rest) flag = takeExpDigits ('-' : acc) rest flag
    takeExp acc rest flag = takeExpDigits acc rest flag

    takeExpDigits acc rest flag =
      let (digits, rest', _) = takeDigits rest
       in (reverse (digits ++ acc), rest', flag)

    takeDigits xs =
      let (d, r) = span isDigit xs
       in (reverse d, r, False)

File: elixir/lib/wildling/json.ex
defmodule Wildling.Json do
  @moduledoc """
  Minimal JSON parser for wildling template files (stdlib only).
  """

  def parse(text) when is_binary(text) do
    {value, rest} = parse_value(skip_ws(text))
    rest = skip_ws(rest)

    if rest != "" do
      raise ArgumentError, "Unexpected trailing JSON content"
    end

    value
  end

  def parse_object(text) do
    value = parse(text)

    if is_map(value) do
      value
    else
      raise ArgumentError, "Template root must be a JSON object"
    end
  end

  defp skip_ws(<<c::utf8, rest::binary>>) when c in [?\s, ?\n, ?\r, ?\t], do: skip_ws(rest)
  defp skip_ws(text), do: text

  defp parse_value(<<"{"::binary, _::binary>> = text), do: parse_object_value(text)
  defp parse_value(<<"["::binary, _::binary>> = text), do: parse_array(text)
  defp parse_value(<<"\""::binary, _::binary>> = text), do: parse_string(text)
  defp parse_value(<<"true", rest::binary>>), do: {true, rest}
  defp parse_value(<<"false", rest::binary>>), do: {false, rest}
  defp parse_value(<<"null", rest::binary>>), do: {nil, rest}

  defp parse_value(<<c::utf8, _::binary>> = text) when c == ?- or (c >= ?0 and c <= ?9) do
    parse_number(text)
  end

  defp parse_value(""), do: raise(ArgumentError, "Unexpected end of JSON")
  defp parse_value(_), do: raise(ArgumentError, "Unexpected character in JSON")

  defp parse_string(<<"\"", rest::binary>>), do: do_parse_string(rest, [])

  defp do_parse_string(<<"\"", rest::binary>>, acc) do
    {acc |> Enum.reverse() |> IO.iodata_to_binary(), rest}
  end

  defp do_parse_string(<<"\\", esc::utf8, rest::binary>>, acc) do
    case esc do
      ?" -> do_parse_string(rest, [?" | acc])
      ?\\ -> do_parse_string(rest, [?\\ | acc])
      ?/ -> do_parse_string(rest, [?/ | acc])
      ?b -> do_parse_string(rest, [?\b | acc])
      ?f -> do_parse_string(rest, [?\f | acc])
      ?n -> do_parse_string(rest, [?\n | acc])
      ?r -> do_parse_string(rest, [?\r | acc])
      ?t -> do_parse_string(rest, [?\t | acc])
      ?u ->
        case rest do
          <<a, b, c, d, more::binary>> ->
            hex = <<a, b, c, d>>

            if Regex.match?(~r/\A[0-9a-fA-F]{4}\z/, hex) do
              code = String.to_integer(hex, 16)
              do_parse_string(more, [<<code::utf8>> | acc])
            else
              raise ArgumentError, "Invalid unicode escape"
            end

          _ ->
            raise ArgumentError, "Invalid unicode escape"
        end

      _ ->
        raise ArgumentError, "Invalid escape"
    end
  end

  defp do_parse_string(<<c::utf8, rest::binary>>, acc) do
    do_parse_string(rest, [<<c::utf8>> | acc])
  end

  defp do_parse_string("", _), do: raise(ArgumentError, "Unterminated string")

  defp parse_number(text) do
    {raw, rest, is_float} = take_number(text, [], false)

    value =
      if is_float do
        case Float.parse(raw) do
          {f, ""} -> f
          _ -> raise ArgumentError, "Invalid number"
        end
      else
        String.to_integer(raw)
      end

    {value, rest}
  end

  defp take_number(<<"-", rest::binary>>, acc, is_float) do
    take_number(rest, ["-" | acc], is_float)
  end

  defp take_number(<<c::utf8, rest::binary>>, acc, is_float) when c >= ?0 and c <= ?9 do
    take_number(rest, [<<c::utf8>> | acc], is_float)
  end

  defp take_number(<<".", rest::binary>>, acc, _) do
    take_number_frac(rest, ["." | acc], true)
  end

  defp take_number(<<c::utf8, rest::binary>>, acc, _is_float) when c in [?e, ?E] do
    take_number_exp(rest, [<<c::utf8>> | acc], true)
  end

  defp take_number(rest, acc, is_float) do
    {acc |> Enum.reverse() |> IO.iodata_to_binary(), rest, is_float}
  end

  defp take_number_frac(<<c::utf8, rest::binary>>, acc, _) when c >= ?0 and c <= ?9 do
    take_number_frac(rest, [<<c::utf8>> | acc], true)
  end

  defp take_number_frac(<<c::utf8, rest::binary>>, acc, _) when c in [?e, ?E] do
    take_number_exp(rest, [<<c::utf8>> | acc], true)
  end

  defp take_number_frac(rest, acc, is_float) do
    {acc |> Enum.reverse() |> IO.iodata_to_binary(), rest, is_float}
  end

  defp take_number_exp(<<"+", rest::binary>>, acc, _) do
    take_number_exp_digits(rest, ["+" | acc], true)
  end

  defp take_number_exp(<<"-", rest::binary>>, acc, _) do
    take_number_exp_digits(rest, ["-" | acc], true)
  end

  defp take_number_exp(rest, acc, is_float) do
    take_number_exp_digits(rest, acc, is_float)
  end

  defp take_number_exp_digits(<<c::utf8, rest::binary>>, acc, _) when c >= ?0 and c <= ?9 do
    take_number_exp_digits(rest, [<<c::utf8>> | acc], true)
  end

  defp take_number_exp_digits(rest, acc, is_float) do
    {acc |> Enum.reverse() |> IO.iodata_to_binary(), rest, is_float}
  end

  defp parse_array(<<"[", rest::binary>>) do
    rest = skip_ws(rest)

    if match?(<<"]", _::binary>>, rest) do
      <<"]", more::binary>> = rest
      {[], more}
    else
      parse_array_items(rest, [])
    end
  end

  defp parse_array_items(text, acc) do
    {value, rest} = parse_value(skip_ws(text))
    rest = skip_ws(rest)

    case rest do
      <<"]", more::binary>> ->
        {Enum.reverse([value | acc]), more}

      <<",", more::binary>> ->
        parse_array_items(more, [value | acc])

      _ ->
        raise ArgumentError, "Expected ',' or ']' in JSON array"
    end
  end

  defp parse_object_value(<<"{", rest::binary>>) do
    rest = skip_ws(rest)

    if match?(<<"}", _::binary>>, rest) do
      <<"}", more::binary>> = rest
      {%{}, more}
    else
      parse_object_pairs(rest, [])
    end
  end

  defp parse_object_pairs(text, acc) do
    {key, rest} = parse_string(skip_ws(text))
    rest = skip_ws(rest)

    case rest do
      <<":", more::binary>> ->
        {value, rest} = parse_value(skip_ws(more))
        rest = skip_ws(rest)
        acc = [{key, value} | acc]

        case rest do
          <<"}", more::binary>> ->
            {Map.new(Enum.reverse(acc)), more}

          <<",", more::binary>> ->
            parse_object_pairs(more, acc)

          _ ->
            raise ArgumentError, "Expected ',' or '}' in JSON object"
        end

      _ ->
        raise ArgumentError, "Expected ':' in JSON object"
    end
  end
end
